### PR TITLE
types: generic constructor

### PR DESCRIPTION
**Pull request by ChristianMurphy**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -4,9 +4,11 @@ import inherits from 'inherits'
  * Create a custom constructor which can be modified without affecting the
  * original class.
  *
- * @template {Function} Class
+ * @template Instance
+ * @template Args
+ * @template {{ new(...Args): Instance }} Class
```
Comment (ChristianMurphy) [inline index.js]: These reflect the arguments that the constructor accepts and the instance type it returns

File: index.js [hunk 2]
```diff
@@ -4,9 +4,11 @@ import inherits from 'inherits'
...
  * @param {Class} Super
- * @returns {Class}
+ * @return {(new(...Args) => Instance) & ((...Args) => Instance)}
```
Comment (ChristianMurphy) [inline index.js]: returning class directly appears to run into https://stackoverflow.com/questions/17382143/create-a-new-object-from-type-parameter-in-generic-class
This overcomes this a constructor and factory are declared and use the reflected constructor and instance information

File: test.js [hunk 3]
```diff
@@ -32,8 +33,7 @@ test('unherit(Super)', function (t) {
 
   var B = unherit(A)
 
-  // eslint-disable-next-line new-cap
-  var b = B('foo', 'bar', 'baz')
+  var b = new B('foo', 'bar', 'baz')
```
Comment (ChristianMurphy) [inline test.js]: this could alternatively be TS ignored

File: index.js [hunk 4]
```diff
@@ -4,9 +4,9 @@ import inherits from 'inherits'
  * Create a custom constructor which can be modified without affecting the
  * original class.
  *
- * @template {Function} Class
+ * @template {{ new(...args: any): any }} Class
  * @param {Class} Super
- * @returns {Class}
+ * @return {new(...args: ConstructorParameters<Class>) => InstanceType<Class>}
```
Comment (ChristianMurphy) [inline index.js]: TypeScript generics appear to strip constructors, this recreates the constructor through generic reflection

File: index.js [hunk 5]
```diff
@@ -45,7 +45,8 @@ export function unherit(Super) {
   /**
    * Constructor accepting variadic arguments.
    *
-   * @returns {any}
+   * @class
+   * @this {Class}
```
Comment (wooorm) [inline index.js]: I guess `this` is *potentially* an instance here? So should it use `Class`? 🤔

File: test.js (hunk 3 above)
Comment (wooorm) [reply test.js]: this *might* be testing when the class is used w/o `new`, and then it’s instantiated by the instanceof check in `Of`

File: index.js (hunk 5 above)
Comment (ChristianMurphy) [reply index.js]: Maybe :man_shrugging: 
Updated this, give the new version a look

**Comment by wooorm** [comment]
Beautiful!

**Review by wooorm — APPROVED** [review]

**Comment by wooorm** [comment]
✨

**Pull request merged**